src/decoder/image.rs

```diff
@@ -447,6 +449,25 @@ impl Image {
 
                 Box::new(Cursor::new(data))
             }
+            CompressionMethod::Fax4 => {
+                let width = u16::try_from(self.width)?;
+                let height = u16::try_from(self.height)?;
+                let mut out: Vec<u8> = Vec::with_capacity(usize::from(width) * usize::from(height));
+
+                let mut buffer = Vec::with_capacity(usize::try_from(compressed_length)?);
+                reader.take(compressed_length).read_to_end(&mut buffer)?;
+
+                // all extant tiff/fax4 decoders I've found always assume that the photometric interpretation
+                // is `WhiteIsZero`, ignoring the tag. ImageMagick appears to generate fax4-encoded tiffs
+                // with the tag incorrectly set to `BlackIsZero`.
+                fax::decoder::decode_g4(buffer.into_iter(), width, Some(height), |transitions| {
```

Comment:
It'd be `reader.take(compressed_length).bytes().map(|b| b.unwrap())` instead. Is that OK with you, in terms of 
* performance impact: you'd get a branch for every single byte
* error handling: you'd get a panic instead of an `Err`, and I (a Rust neophyte) don't think there's a way around that

Reply:
The bit reader in the fax decoder pulls single bytes anway. I started changing the api to accept an Iterator of Result<u8, E>. (so `reader.take(compressed_bytes).bytes()` will work)

Reply:
You could also see if taking a `BufRead` impl would improve performance over having to call into the reader for each individual byte. We're currently in the process of converting various parts of the _image-rs_ APIs to use those over normal `Read` impls.

Reply:
FWIW, I do not think it is worth worrying too much about micro-optimizations here. It makes sense to handle malicious inputs without blowing up the heap, but in practice all of the images I'm seeing are crappy low-res scans of checks from god-knows-what device a regional bank purchased in 1993. Really doubt there are many images of this format in the wild that demand fully optimized best-case performance